osutil/disks/disks.go

```diff
@@ -0,0 +1,362 @@
+	}
+
+	if opts != nil && opts.IsDecryptedDevice {
+		// verify that the mount point is indeed a mapper device, it should:
+		// 1. have DEVTYPE == disk from udev
+		// 2. have dm files in the sysfs entry for the maj:min of the device
+		if props["DEVTYPE"] != "disk" {
+			// not a decrypted device
+			return nil, fmt.Errorf("mountpoint source %s is not a decrypted device", mountpointSrc)
+		}
+
+		// TODO:UC20: these files should also be readable through udev env
+		//            properties, but currently aren't available for some reason
+		dmUUID, err := ioutil.ReadFile(filepath.Join(devBlockDir, d.Dev(), "dm", "uuid"))
+		if err != nil && os.IsNotExist(err) {
+			return nil, fmt.Errorf("mountpoint source %s is not a decrypted device", mountpointSrc)
+		}
+
+		dmName, err := ioutil.ReadFile(filepath.Join(devBlockDir, d.Dev(), "dm", "name"))
+		if err != nil && os.IsNotExist(err) {
+			return nil, fmt.Errorf("mountpoint source %s is not a decrypted device", mountpointSrc)
+		}
+
+		// trim the suffix of the dm name from the dm uuid to safely match the
+		// regex - the dm uuid contains the dm name, and the dm name is user
+		// controlled, so we want to remove that and just use the luks pattern
+		// to match the device uuid
+		// we are extra safe here since the dm name could be hypothetically user
+		// controlled via an external USB disk with LVM partition names, etc.
+		dmUUIDSafe := bytes.TrimSuffix(
+			bytes.TrimSpace(dmUUID),
+			append([]byte("-"), bytes.TrimSpace(dmName)...),
+		)
+		matches := luksUUIDPatternRe.FindSubmatch(dmUUIDSafe)
+		if len(matches) != 2 {
+			// the format of the uuid is different - different luks version maybe?
+			return nil, fmt.Errorf("cannot verify disk: partition %s does not have a valid luks uuid format", d.Dev())
+		}
+
+		// the uuid is the first and only submatch, but it is not in the same
+		// format exactly as we want to use, namely it is missing all of the "-"
+		// characters in a typical uuid, i.e. it is of the form:
+		// ae6e79de00a9406f80ee64ba7c1966bb but we want it to be like:
+		// ae6e79de-00a9-406f-80ee-64ba7c1966bb so we need to add in 4 "-"
+		// characters
+		fullUUID := string(matches[1])
+		realUUID := fmt.Sprintf(
+			"%s-%s-%s-%s-%s",
+			fullUUID[0:8],
+			fullUUID[8:12],
+			fullUUID[12:16],
+			fullUUID[16:20],
+			fullUUID[20:],
+		)
+
+		// now finally, we need to use this uuid, which is the device uuid of
+		// the actual physical encrypted partition to get the path, which will
+		// be something like /dev/vda4, etc.
+		byUUIDPath := filepath.Join("/dev/disk/by-uuid", realUUID)
+		props, err = udevProperties(byUUIDPath)
+		if err != nil {
+			return nil, fmt.Errorf("cannot get udev properties for encrypted partition %s: %v", byUUIDPath, err)
+		}
+
+		// after this, the rest of the function is the same, with props
+		// "redirected" to the physical partition
+		if devType := props["DEVTYPE"]; devType != "partition" {
```

Comment:
I don't think we need this check here, for consistency we should be ok again if the encrypted disk is a disk (and not a partition), like for the non decrypted case.

Reply:
I am worried about this code doing too much in cases we don't yet understand. I would rather we explicitly fail in cases we don't know what is happening here rather than try and have this code be more general. Remember that if this code mounts the wrong thing we are exposing ourselves to a dangerous bug where an external user controlled rootfs is mounted when we have already decrypted the actual encrypted data partition.

Reply:
I don't quite agree about the "don't understand", anyway, what we can do is have a method Partioned() or HasPartitions() and a corresponding flag in disk, and set it to true in the code path that goes through finding ID_PART_ENTRY_DISK  (we could also double check that DEVTYPE is partition in that path). At that point the code using this and/or FindMatchingPartitionUUID could bail out early if the flag is false.

Reply:
re: "don't understand" I mean exactly that. I don't understand how likely or easy it is to have a device mapper mounted at a mount point, resolve back to something encrypted which is not a physical partition. I am not a systemd expert insofar as I cannot confidently claim that allowing this code to silently continue to work, returning the major/minor for something that we don't know what it is during the initramfs, does not lead to security issues.

I would much rather expand this code to work in more general cases when we need it to and understand what environment it is being called from. I guess in general I'm not really comfortable with the scope of this package growing so much since it's not used anywhere else and so trying to make it generic for unknown use cases when it also needs to be robust and secure in the initramfs are kind of at odds it seems.

Reply:
also re: your proposal on having some kind of flag in disk, I'm not sure that does all of what we want either, because for example should `MountPointIsFromDisk` also check this flag? The name of the function does not imply that it should fail because it is partitioned, but if it does not then we still have this silent opportunity for something we mounted and decrypted to end up being not a partition, but have matching major minor numbers if I understand correctly. I.e. what about this case:

1. we get ubuntu-boot or ubuntu-seed partition mounted somehow
2. we get the disk from ubuntu-boot or ubuntu-seed via DiskFromMountPoint
3. we use that disk to see if /run/mnt/data is from the same disk, this calls MountPointIsFromDisk with opts.IsDecrypted == true

Hence, in 3) if /run/mnt/data somehow came from something that is not an encrypted partition, what should MountPointIsFromDisk do? Should it assert that both *disk's have partitions and also have the same device major/minor? I don't know what other context that MountPointIsFromDisk might be used in, but the function name does not imply that the mountpoint's source needs to have DEVTYPE==partition.

Reply:
we check that the flag has the same value, it should, but the paranoia doesn't cost us anything. Our problem here is that this code is too much code to live as a helper in snap-bootstrap, but if we make it too specific I'm sure it will confuse future readers. I'm not trying to make it super generic, I'm trying though to make it enough generic as not to be confusing and do add enough extra checks to make it robust enough.

Reply:
alright this is implemented now with some more unit tests